### PR TITLE
Fixes: Closes issues #12, #16, #17

### PR DESCRIPTION
**Pull request by johnsonjh**
* Closes:
  *  #12
  *  #16
  *  #17

> Did this on my phone, please review and make sure I'm not being dumb.

### CODE REVIEW

**Comment by johnsonjh** [comment]
Should be good to go

File: sirfilecache.c
Comment (aremmell) [inline sirfilecache.c:1]: Here's the trick: _sir_safefree sets the pointer to NULL after it frees; so this return NULL is redundant. I think clang-analyzer missed that.

File: sirfilesystem.c
Comment (aremmell) [inline sirfilesystem.c:1]: Well, this isn't the only place `_sir_getappfilename` is called (it's called all over). Why is freeing it inside `if (!resolved)` good enough? I'm confused about that.

File: tests/tests.c
Comment (aremmell) [inline tests/tests.c:1]: I believe this should be size_t, no? And how does declaring `err` separately resolve an issue? Sorry, I'm confused about that too.

File: sirfilecache.c
Comment (johnsonjh) [reply sirfilecache.c:1]: Yeah, it's a bit strange ... I just checked Clang and also Purify and they both complain if it isn't there, so I'd say leave it, can't hurt.

File: tests/tests.c
Comment (johnsonjh) [reply tests/tests.c:1]: The root of the problem was the previous typo which is now fixed, so this shouldn't be needed to pass now; I'll re-simplify it.

File: sirfilesystem.c
Comment (johnsonjh) [reply sirfilesystem.c:1]: Reversing this patch on this branch, i.e.:
```diff
diff --git a/sirfilesystem.c b/sirfilesystem.c
index e840bf3..19bb927 100644
--- a/sirfilesystem.c
+++ b/sirfilesystem.c
@@ -239,0 +240 @@ char* _sir_getappfilename(void) {
+       _sir_safefree(buffer);
@@ -248,2 +249 @@ char* _sir_getappbasename(void) {
-    if (!_sir_validstr(filename)) {
-       _sir_safefree(filename);
+    if (!_sir_validstr(filename))
@@ -251 +250,0 @@ char* _sir_getappbasename(void) {
-    }
```

Analzyer reports the following two errors with these paths:

[IMG]
[IMG]

Comment (johnsonjh) [reply sirfilesystem.c:1]: @aremmell 

All other things you noted are addressed.

**Comment by johnsonjh** [comment]
Rebased; fixes all remaining scan-build errors and passes all tests.

**Pull request merged**